src/lfortran/semantics/ast_common_visitor.h

```diff
@@ -1337,10 +1337,38 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         for (size_t i=0; i<n_dim; i++) {
             ASR::dimension_t dim; dim.m_length = nullptr; dim.m_start = nullptr;
             dim.loc = m_dim[i].loc;
-            if (m_dim[i].m_start) {
+           if (m_dim[i].m_start) {
                 this->visit_expr(*m_dim[i].m_start);
                 dim.m_start = ASRUtils::EXPR(tmp);
-            } else {
+                ASR::expr_t* start = dim.m_start;
+                if (ASR::is_a<ASR::Var_t>(*start)) {
+
+                    ASR::Var_t* start_var = ASR::down_cast<ASR::Var_t>(start);
+                    ASR::symbol_t* start_sym = start_var->m_v;
+                    SymbolTable* symbol_scope = ASRUtils::symbol_parent_symtab(start_sym);
+                    if (ASR::is_a<ASR::Variable_t>(*start_sym)) {
+                        ASR::Variable_t* start_variable = ASR::down_cast<ASR::Variable_t>(start_sym);
+
+                        if (start_variable->m_storage != ASR::storage_typeType::Parameter) {
+
+                            if(!in_Subroutine && (symbol_scope->counter == current_scope->counter)) {
+                                throw SemanticError("Expecting a scalar integer or parameter annotated integer variable ",m_dim[i].m_start->base.loc);
+                            }
+                        }
+                    }
+                    else {
```

Comment:
Use this syntax:
```suggestion
                    } else {
```
The same at other places below.